### PR TITLE
Extract open issues and organize Expressions section

### PR DESCRIPTION
**Pull request by jcouv**
Relates to test plan https://github.com/dotnet/roslyn/issues/66722

### CODE REVIEW

File: proposals/extensions.md [hunk 1]
```diff
@@ -601,41 +760,61 @@ The preceding rules mean:
 - and that extension methods declared directly in a namespace take precedence
   over extension methods imported into that same namespace with a using namespace directive.
 
-TODO(static) clarify behavior for extension on `object` or `dynamic` used as `dynamic.M()`?
```
Comment (jcouv) [inline proposals/extensions.md:604]: :memo: This is converted to a test with a PROTOTYPE comment about diagnostic quality in [PR](https://github.com/dotnet/roslyn/pull/73994)

**Review by MadsTorgersen — APPROVED** [review]
This is a good restructuring! Makes it a lot easier to get an overview on the current state of design, and which are the most pressing issues.

File: proposals/extensions.md [hunk 2]
```diff
@@ -116,6 +109,254 @@ public implicit extension E for S
...
+  - `Add` in collection initializers
+  - `GetPinnableReference` in fixed statements
+  - `GetAwaiter` in `await` expressions
+  - `Dispose` in `using` statements
+
+## Open issue: need to disallow extensions within interfaces with variant type parameters  
+
+Will need to check with the WG as I don't recall the reasoning for this.
+
+## Open issue: need to specify the type erasure design
```
Comment (jjonescz) [inline proposals/extensions.md:341]: What's the difference between this and "Open issue: need to specify type erasure" above?

File: proposals/extensions.md [hunk 3]
```diff
@@ -632,41 +855,62 @@ The preceding rules mean:
...
-TL;DR: If no candidate is applicable, then we attempt extension indexer access instead.
+TL;DR: For non-extension types, we'll fall back to an implicit extension member lookup. For extension types, we include indexers from the underlying type.
 
 We modify the [element access rules](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#128111-general) as follows:
 
-/[...]
 An *element_access* is dynamically bound if \[...]
 If the *primary_no_array_creation_expression* of an *element_access* is a value of an *array_type*, the *element_access* is an array access. 
-Otherwise, the *primary_no_array_creation_expression* shall be a variable or value of a class, struct, or interface type  TODOTODO
+Otherwise, the *primary_no_array_creation_expression* shall be a variable or value of a class, struct, interface, /***or extension** type
```
Comment (jjonescz) [inline proposals/extensions.md:866]: Did you mean to write `\*` instead of `/*` ? But I don't know why would that be needed, this seems to work fine:

```suggestion
Otherwise, the *primary_no_array_creation_expression* shall be a variable or value of a class, struct, interface, **or extension** type
```

Applies to other similar changes throughout the doc - I see usages of `/*` and `/[` that were probably meant to be `\*` and `\[` (although the escaping might be also unnecessary).

Comment (jcouv) [reply proposals/extensions.md]: Tried again and it looks like the escape sequence is no longer needed. Removed :-)

File: proposals/extensions.md (hunk 2 above)
Comment (jcouv) [reply proposals/extensions.md]: Merged now. Thanks

File: proposals/extensions.md [hunk 4]
```diff
@@ -500,7 +497,7 @@ We modify the [accessibility constraints](https://github.com/dotnet/csharpstanda
 
 The following accessibility constraints exist:
 - [...]
-- \***The underlying type of an extension type shall be at least as accessible as the extension type itself.**
+- ***The underlying type of an extension type shall be at least as accessible as the extension type itself.**
```
Comment (jjonescz) [inline proposals/extensions.md:500]: Nit: it seems excessive to use three `*` - the first is rendered as a literal `*` and the remaining two make the text bold:

image

Comment (jcouv) [reply proposals/extensions.md:500]: This is intended. If found that the bold is not always easy to see, so the `*` helps spot additions even when the bold isn't.
It looks like GitHub may have adjusted their fonts and the bold seems easier to see than before, but I think it's better to keep for now. It'll be easy to remove if we decide this extra formatting isn't necessary.

**Comment by jcouv** [comment]
@AlekseyTs for review. Thanks

File: proposals/extensions.md [hunk 5]
```diff
@@ -116,6 +109,251 @@ public implicit extension E for S
...
+
+static class Extensions
+{
+    public static void M(this C c) { }
+}
+```
+
+## Open issue: removing zero-arity-matches-any from var scenario
+
+By the current rules, there are some unfortunate interactions between member access and natural function type.
```
Comment (AlekseyTs) [inline proposals/extensions.md:134]: >there are some unfortunate interactions between member access and natural function type [](http://example.com/codeflow?start=22&length=87)

This is not specific. What interactions? Why might one consider them unfortunate?

File: proposals/extensions.md [hunk 6]
```diff
@@ -116,6 +109,251 @@ public implicit extension E for S
...
+> Note: When the result of such a member lookup is a method group and K is zero,
```
Comment (AlekseyTs) [inline proposals/extensions.md:135]: It is not obvious what is the context for the quote (where it is taken from) and how it is relevant to the issue.

File: proposals/extensions.md [hunk 7]
```diff
@@ -116,6 +109,251 @@ public implicit extension E for S
...
+> the method group can contain methods having type parameters.
+> This allows such methods to be considered for type argument inferencing. end note
+
+I would like to brainstorm tweaks to member access and natural function type to make this work better.
```
Comment (AlekseyTs) [inline proposals/extensions.md:139]: Better is a subjective term, consider describing the behavior that you think is better.

File: proposals/extensions.md [hunk 8]
```diff
@@ -116,6 +109,251 @@ public implicit extension E for S
...
+
+```csharp
+var x = C.Member; // error: member lookup finds C.Member (method group) and lacks type arguments to apply to that match
```
Comment (AlekseyTs) [inline proposals/extensions.md:142]: It might help including what does the error say exactly

File: proposals/extensions.md [hunk 9]
```diff
@@ -116,6 +109,251 @@ public implicit extension E for S
...
+
+## Open issue: what kind of type is an extension type?
+
+Many sections of the spec need to consider the kind of type. Consider a few examples:  
+- the spec for a conditional element access `P?[A]B`
+considers whether `P` is a nullable value type. So it will need to handle the case
+where `P` is an instance of an extension type on a nullable value type,
+- the spec for an object creation considers whether the type is
+a value_type, a type_parameter, a class_type or a struct_type,
+- the spec for satisfying constraints also consider what kind of type were dealing with.
```
Comment (AlekseyTs) [inline proposals/extensions.md:195]: Typo? "what kind of type **were** dealing with"

File: proposals/extensions.md [hunk 10]
```diff
@@ -116,6 +109,251 @@ public implicit extension E for S
...
+
+It may be possible to address all those cases without changing each such section of the spec,
+but rather by adding general rules ("an extension on a class type is considered a class type" or some such).
+
+## Open issue: need to specify type erasure
+
+It should be done in a way that allows switching between the extension type and the underlying type without binary break.  
+It should allow for using type parameters as type parameters: `void M<T>(E<T> e)`.
```
Comment (AlekseyTs) [inline proposals/extensions.md:203]: Did you mean to say: "It should allow for using type parameters as type **arguments**" ?

File: proposals/extensions.md [hunk 11]
```diff
@@ -116,6 +109,251 @@ public implicit extension E for S
...
+## Open issue: need to specify type erasure
+
+It should be done in a way that allows switching between the extension type and the underlying type without binary break.  
+It should allow for using type parameters as type parameters: `void M<T>(E<T> e)`.  
+It should allow for using extension types as type arguments without violating type parameter constraints: `C<E>` with `class C<T> where T : struct, I { }`.  
+
+The current proposal is to use an attribute with a string representing the type with extensions un-erased.
+For example: `void M(E)` would be emitted as `void M([Attribute("E")] UnderlyingType)`.
+
+The serialization format would be the same as the one used for `typeof` in attributes, but with support for type parameters (using `T!` and `T!!` syntax).
```
Comment (AlekseyTs) [inline proposals/extensions.md:209]: I thought the proposal was to follow the way specified in ECMA-335 (i.e. to refer by ordinal, rather than by name):
>II.9 Generics
>Within a generic type definition, its generic parameters are referred to by their index. Generic
parameter zero is referred to as !0, generic parameter one as !1, and so on. Similarly, within the body
of a generic method definition, its generic parameters are referred to by their index; generic parameter
zero is referred to as !!0, generic parameter one as !!1, and so on.

File: proposals/extensions.md [hunk 12]
```diff
@@ -473,45 +669,73 @@ We modify the [member access rules](https://github.com/dotnet/csharpstandard/blo
...
-2. indexer access scenarios where the set of *applicable* candidate indexers is empty.
-3. TODO(static) there may be more scenarios (operator resolution, delegate conversion, natural function types)
 
-That is covered below.
-
-TODO3(static) Is the "where `T` is not a type parameter" portion still relevant?
+Note: We allow static lookups on type parameters only for members that are static virtual members. 
+Since extensions members are not virtual, we don't allow static extension lookups on type parameters either.
+```
+void M<T>(T t) where T
```
Comment (AlekseyTs) [inline proposals/extensions.md:695]: The constraint is not finished. Did you mean "where T : U"?

File: proposals/extensions.md [hunk 13]
```diff
@@ -715,6 +956,30 @@ The search proceeds as follows:
...
+#### This access
+
+We modify the [this access rules](https://github.com/dotnet/csharpstandard/blob/standard-v7/standard/expressions.md#12813-this-access) as follows:
+
+A this_access has one of the following meanings:
+- [...]
+- ***When this is used in a primary_expression within an instance method or instance accessor of an extension with a class underlying type,
+  it is classified as a value. The type of the value is the instance type of the extension within which the usage occurs, 
+  and the value is a reference to the object for which the method or accessor was invoked.**
+- ***When this is used in a primary_expression within an instance method or instance accessor of an extension with a struct underlying type,
```
Comment (AlekseyTs) [inline proposals/extensions.md:968]: Need to specify the meaning for a scenario when the underlying type is a type parameter not known to be a reference type or a value type.

File: proposals/extensions.md [hunk 14]
```diff
@@ -715,6 +956,30 @@ The search proceeds as follows:
 - If no extension indexer is found to be suitable for the element access
   in any enclosing scope, a compile-time error occurs.
 
+#### This access
+
+We modify the [this access rules](https://github.com/dotnet/csharpstandard/blob/standard-v7/standard/expressions.md#12813-this-access) as follows:
+
+A this_access has one of the following meanings:
+- [...]
+- ***When this is used in a primary_expression within an instance method or instance accessor of an extension with a class underlying type,
```
Comment (AlekseyTs) [inline proposals/extensions.md:965]: > with a class underlying type

A class underlying type, or an underlying type known to be a reference type? For example, an interface type is not a class and is not a struct, and is not a type parameter.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 10)

File: proposals/extensions.md (hunk 5 above)
Comment (jcouv) [reply proposals/extensions.md:134]: See the example below (line 142). A member that cannot work in the end gets in the way. We've taken steps in overload resolution and more recently in natural type to reduce such behaviors (discard bad candidates earlier).
I'm not sure we should do something here, maybe the current behavior is desirable fallout of current rules...

Comment (jcouv) [reply proposals/extensions.md:134]: I'm going to remove this open issue. I don't have a good proposal for how to make this scenario work and I'm no longer convinced it should.

Here's an updated summary of the question for the record:

## Open issue: removing zero-arity-matches-any from var scenario

By the current rules, a generic method which cannot possibly work in a function type scenario can hide
a member that would work:

```csharp
var x = C.Member; // error: member lookup finds C.Member (method group) and lacks type arguments to apply to that match

class C
{
    public static void Member<T>() { }
}

implicit extension E for C
{
    public static int Member = 42;
}
```

Is this a scenario we think should work?

For context, from the [member access rules](https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/expressions.md#12871-general),
we perform a member lookup which includes generic methods even when no type arguments are present:
> [...] if a member lookup of I in E with K type parameters produces a match, then E.I is evaluated and classified as follows
> Note: When the result of such a member lookup is a method group and K is zero,
> the method group can contain methods having type parameters.
> This allows such methods to be considered for type argument inferencing. end note

See details from the [member lookup rules](https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/expressions.md#1251-general):

> When K is zero, methods having type parameters are not removed, since the type inference process might be able to infer the type arguments.

See details on [natural type of method groups](https://github.com/dotnet/csharplang/blob/main/proposals/method-group-natural-type-improvements.md)

File: proposals/extensions.md (hunk 13 above)
Comment (jcouv) [reply proposals/extensions.md:968]: Yes. I meant to have this as an open issue but forgot. Added.

**Comment by jcouv** [comment]
@AlekseyTs Please take another look. Thanks

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 11)
